libcoap/src/message/mod.rs

```diff
@@ -475,14 +471,49 @@ impl CoapMessage {
         coap_pdu_set_type(raw_pdu, self.type_.to_raw_pdu_type());
         coap_pdu_set_code(raw_pdu, self.code.to_raw_pdu_code());
         let message = self.as_message_mut();
-        let token: &[u8] = message.token.as_ref().ok_or(MessageConversionError::MissingToken)?;
-        if coap_add_token(raw_pdu, token.len(), token.as_ptr()) == 0 {
-            return Err(MessageConversionError::Unknown);
+        let cur_token = coap_pdu_get_token(raw_pdu);
+
+        let cur_token = (cur_token.length > 0).then(|| core::slice::from_raw_parts(cur_token.s, cur_token.length));
+        // Return error if no token was supplied.
+        if cur_token.is_none() && message.token.is_none() {
+            return Err(MessageConversionError::MissingToken);
         }
+        // Update token if necessary.
+        if message.token.is_some() && cur_token != message.token.as_deref() {
+            let token = message.token.as_ref().unwrap();
+            if coap_add_token(raw_pdu, token.len(), token.as_ptr()) == 0 {
+                return Err(MessageConversionError::Unknown);
+            }
+        }
```

Comment:
I was wondering here if you could just write the token regardless of the comparison result, as you would need to iterate over the buffers twice when they do not match.